### ISSUE TITLE
replace `length` with `percent_along` & actual edge length in `trace_attributes`

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4192]
Currently `length` of `trace_attributes` is referring to `percent_along` * full edge length, i.e. the start & end edge. For live traffic matching (in particular setting breakpoints), it'd be much nicer to have the full edge length and its percent_along.

Whenever I implement live traffic I keep coming back to this and so far I patched it. We could call it `full_length` so to not cause trouble for anyone?

**Comment by nilsnolde** [comment]
No sorry, `percent_along` is no good in this context.. Actually simply having a `total_length` for each edge alongside the current `length` is sufficient, as we'd only be interested in the percentage of the edge covered to determine the breakpoint(s) (unless it's single-edge routes, but that can be handled differently).

**Comment by kevinkreiser** [comment]
to do the traffic breakpoints i think you will need percents along a given edge. the reason for this is because a match can, as your are implying be only part of an edge but you cant know which part. a match may only run from 25% from the beginning to 75% from the beginning and then stop. if you had the match length for the edge and the total length you'd know that the match only used 50% of the edge but you wouldnt know that its directly in the middle of the edge with 25% on either side of the match unused.

also because this information will tend to mostly be 0% to 100% for a lot of the non-degenerate cases i think we should only include this information if its a partially matched edge, that is it is adjacent to a match point that "begins/end a discontinuity" (even the ones at the start and end of the entire request shape, technically not discontinuities per se i guess but still).

also its possible i misunderstood what you are saying :smile:

**Comment by nilsnolde** [comment]
From the view of live traffic, your case is a trivial route, i.e. only one edge, from 25% to 75%. Not sure how others do it, but I catch that before it goes to routing/map-matching by `/locate`ing LRPs. That's what I meant with

> unless it's single-edge routes which can have 2 breakpoints, but that can be handled differently

And then I guess having both, the covered `length` and the `total_length` of the edge would be fine, as it could only be the first/last edge, which the code knows at that point.

If we'd do the percent_along instead, we'd have to do a `source_percent_along` and a `target_percent_along`, which works for me as well and is maybe more useful generally, apart from live traffic stuff?

**Comment by kevinkreiser** [comment]
im not just talkign about trivial ones im talking about ones that match a little stop matching match a little stop matching for a whole trace, could be miles of gps points. we have to remember that the trace attributes api isn't one use-case (eg traffic matching), you can feed a big long shape into it and it can have partial matches at the beginning and end and also in the middle.

if we want to return useful information about what segment of an edge was partially matched we have to be robust to all the use-cases that we'll encounter which is pretty much all of them with the statistical nature of map matching. just to be clear with a visual a map match can be like:


```
graph: o-----------------o---------------------o---------------o------------o
match: XXXX++++++++++XXXXXXX++++XXXXXXXXXXXXXXXXXXX++++++++++++++++++XXXXXXXX
```

where you the `+`'s are matched and the `X`'s are not. in this case we have a bunch of discontinuities and need to give percentages to make any sense of it (at least i think so without going to review the api docs too much haha)

**Comment by nilsnolde** [comment]
Yeah, I was just focusing on my own little use case with the example of a trivial route. But thanks for the wider view.

In the end I guess we'll need both percent_alongs for each partial edge and for each edge the total_length. Would you agree with that? The current `length` is pretty ambiguous and while helpful for lots of cases, the other properties put the user in full control. For backwards compat I'd guess we don't want to either re-purpose the `length` or even remove it, so then:
- add `length` and `full_length` for **each edge**  (when the attributes controller asks for `edge.length`)
- add `source_percent_along` & `target_percent_along` for **partial edges only** (IMO it's ok to also guard this with the controller's `edge.length` since that's its only use?). Even if a partial match is only one-side-partial (starts in the middle but continues to the end, or vice versa), I'd add both percent_alongs for clarity

WDYT?

**Comment by kevinkreiser** [comment]
I really don't think we need the total length of each edge. it will be redundant in all cases. for the non partial edges it's trivially redundant. for the partial edges you can get total length with simple arithmetic
